server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -773,8 +780,77 @@ public synchronized void handleRecoveryFailure(ShardRouting shardRouting, boolea
+            if (indexShard != null
+                && shardRouting.primary() == false
+                && shardRouting.state() == ShardRoutingState.INITIALIZING
+                && indexShard.state() == IndexShardState.POST_RECOVERY) {
+                segmentReplicationTargetService.startReplication(
+                    ReplicationCheckpoint.empty(shardRouting.shardId()),
+                    indexShard,
+                    new SegmentReplicationTargetService.SegmentReplicationListener() {
+                        @Override
+                        public void onReplicationDone(SegmentReplicationState state) {
+                            logger.trace(
+                                () -> new ParameterizedMessage(
+                                    "[shardId {}] [replication id {}] Replication complete, timing data: {}",
+                                    indexShard.shardId().getId(),
+                                    state.getReplicationId(),
+                                    state.getTimingData()
+                                )
+                            );
+                            shardStateAction.shardStarted(
+                                shardRouting,
+                                primaryTerm,
+                                "after " + recoveryState.getRecoverySource(),
+                                SHARD_STATE_ACTION_LISTENER
+                            );
+                        }
+
+                        @Override
+                        public void onReplicationFailure(
+                            SegmentReplicationState state,
+                            ReplicationFailedException e,
+                            boolean sendShardFailure
+                        ) {
+                            logger.trace(
+                                () -> new ParameterizedMessage(
+                                    "[shardId {}] [replication id {}] Replication failed, timing data: {}",
+                                    indexShard.shardId().getId(),
+                                    state.getReplicationId(),
+                                    state.getTimingData()
+                                )
+                            );
+                            if (sendShardFailure == true) {
+                                logger.error("replication failure", e);
+                                indexShard.failShard("replication failure", e);
+                            }
+                            handleRecoveryFailure(shardRouting, sendShardFailure, e);
+                        }
+                    }
+                );
+            } else {
+                shardStateAction.shardStarted(
+                    shardRouting,
+                    primaryTerm,
+                    "after " + recoveryState.getRecoverySource(),
+                    SHARD_STATE_ACTION_LISTENER
+                );
+            }
+        } else {
+            shardStateAction.shardStarted(
```

Comment:
Nit - this is now invoked 3x. You could clean this up by using a StepListener that when completes marks the shard started.

Reply:
Thanks @mch2 sure I can do that